### ISSUE TITLE
Compiler detection is incorrect for native LLVM/Clang install on Windows

### ISSUE THREAD

**Issue by MarkIngramUK** [mesonbuild/meson#2560]
`ClangCompiler.get_soname_args` assumes that Clang on Windows is running under MINGW. It then tries to generate `-Wl,-soname` options, which are incompatible with [Clang 5.0.0.0 running natively on Windows](http://releases.llvm.org/5.0.0/LLVM-5.0.0-win64.exe).

`CompilerArgs.to_native` adds groups via `-Wl,--start-group` and `-Wl,--end-group` (as `get_compiler_uses_gnuld` returns `true` for Clang on Windows).

`CCompiler.gen_import_library_args` adds the `-Wl,--out-implib` option.

`CCompiler.build_rpath_args` adds the `-Wl,-rpath` option.

Related options `-Wl,--as-needed` and `-Wl,--no-undefined` aren't supported either, but they can be disabled via the `b_asneeded` and `b_lundef` flags, respectively.

**Comment by jibsen** [comment]
Native Clang on Windows can target both MSVC and mingw-w64 (I think they started out targeting MinGW, then added support for MSVC, and now MSVC is the default). Clang also supplies clang-cl, a cli that is compatible with MSVC cl.exe.

I've had some issues getting both options to work with CMake on AppVeyor (see for instance the squash [AppVeyor script](https://github.com/quixdb/squash/blob/5928faf1cc81ab47bce93431417bb5ae9b2330bc/.appveyor.yml)).

So I would love if meson were able to support both.

**Comment by xgalaxy** [comment]
Yes. Support for this would be great. Preferably with an option to choose either the GNU compatible (clang/clang++) or the msvc adapter (clang-cl).

**Comment by XVilka** [comment]
Ping?

**Comment by nirbheek** [comment]
> Ping?

That only really works for pull requests, not issues. See also, "help wanted" label. ;)

We'd be happy to review a PR that fixes this and adds a test for it so we can properly support GNU clang (easy) and clang-cl (harder).

**Comment by barcharcraz** [comment]
fwiw, on windows it looks like both "clang" and "clang++" always compile using the msvc abi, the difference is in how they parse arguments and how good they are at figuring out how to link the C library.

You can see this by taking the following program
```cpp
namespace test {
    void testfunction(int t) {

    }
}
int main(int argc, char** argv) {
    test::testfunction(4);
}
```

and compiling it with `clang++ -c test.cpp` on windows. then looking at test.o in dumpbin or similar, if you see `?testfunction@test@@YAXH@Z` then clang is compiling with MSVC ABI

if you compile with `clang++ -c -target x86_64-pc-windows-gnu test.cpp` you should instead see `_ZN4test12testfunctionEi`

**Comment by barcharcraz** [comment]
note that it's likely that if you configure/build clang with gcc on windows or with gcc on unix targeting windows then you'll get a clang.exe that defaults to x86_64-pc-windows-gnu!

**Comment by barcharcraz** [comment]
AND if you have visual studio installed you may have a version of clang that's neither of these and that is the clang parser with C2 (microsoft's backend) output!!!! For all intents this should be similar to clang with x86_64-pc-windows-msvc

**Comment by dcbaker** [comment]
@jon-turney, is this closable now?

**Comment by jon-turney** [comment]
Not yet fixed.  Possibly a duplicate of #4232, though.

**Comment by bonzini** [comment]
"Possible duplicate" closed, original report mostly fixed by commits around 06dcbd50eea47b3182081527ea1c0ada01d4d847. Closing.
